### PR TITLE
DATA-534 Map Saving Localizing (SLAM)

### PR DESCRIPTION
**Pull request by jeremyhyde-viam**
Adding localization mode. Currently this is done in two areas, one by setting the localization flag and also by removing map updates and checks of number of keyframes.

JIRA Ticket: https://viam.atlassian.net/browse/DATA-534

Associated PR: https://github.com/viamrobotics/rdk/pull/1473 and also https://github.com/viamrobotics/rdk/pull/1548

### CODE REVIEW

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (jeremyhyde-viam) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:489]: added this log and print for consistency between offline/online, however its not currently being used in integration tests in offline mode

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.h:64]: [q] Is `n_key_frames` something that is save to be and/or should be a global variable within the slam service? Is `n_key_frames` not used to count the number of keyframes in each submap depending on the map you're looking at? If I'm wrong, what does it stand for?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc:113]: [q] Why does this have to change? I would have assumed that it would have had to be equal to a value of `60` before, too?

**Review by kkufieta — COMMENTED** [review]
A few nits and questions. Another question: did you make changes in the code of `ORB_SLAM3`?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:504]: was referring to the loop above in the online and offline functions but will remove due to confusion

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:489]: Fine with removing the log, as we have the `Passing image to SLAM` log already. Integration tests for offline mode are already setup and only look to see if all images are completed currently

**Review by JohnN193 — APPROVED** [review]
LGTM, but seems like something weird happened to your ORB_SLAM3 repo, would make sure thats up to date before merging

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc
Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc]: in this test no value is given and so it was defaulted to 0. this behavior is not desired now that localization is 0. We can either have it default to extremely high map rate value to avoid saving a map OR treat it like the data rate and throw an error.

Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc]: I'm in favor of option 2

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.h:64]: number of keyframes will change when you have a valid map update occur that introduces a new key frame. This will cause us to grab all the map points again so we have the most up-to-date map.

If delocalization occurs or a new map is transferred to, this will also change the number of keyframes and cause us to grab a the new data. However I did just realize that if the new map has the same number of key frames the grabbing the new map data would not occur. This was a problem with the past method as well so I have added an additional check for if the current map id has changed as well

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc
Comment (kkufieta) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc]: hmn let me think this through: If no `map_rate_sec` is provided via RDK, that means something went wrong, correct? Since RDK always should be sending a value for `map_rate_sec`. If we run orbslam locally, though, and the user doesn't provide a value for `map_rate_sec`, I would assume that the behavior should be the same as if the user doesn't provide a value through the config - so setting it to a default value of `60` would make sense to me, since it is coherent with how we handle it through RDK, and since providing this parameter should be optional. Am I making sense?

Comment (kkufieta) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc]: Discussed this in person - agreed to throw an error

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:750]: [nit] Mention `map_rate_sec` in the error message. Example: `"map_rate_sec value is required"`

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc:111]: [nit] I think we should definitely mention the actual missing variable. In this case, I don't know what the "camera data rate" is referring to. Saying that `data_rate_ms` has to be defined gives clear instructions and takes out the guessing.

**Review by kkufieta — APPROVED** [review]
LGTM once nits are addressed

**Pull request merged**
